web/api/v1/api.go

```diff
@@ -592,37 +592,18 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
 		warnings storage.Warnings
 	)
 	if len(matcherSets) > 0 {
-		hints := &storage.SelectHints{
-			Start: timestamp.FromTime(start),
-			End:   timestamp.FromTime(end),
-			Func:  "series", // There is no series function, this token is used for lookups that don't need samples.
-		}
-
+		var callWarnings storage.Warnings
 		labelValuesSet := make(map[string]struct{})
-		// Get all series which match matchers.
-		for _, mset := range matcherSets {
-			s := q.Select(false, hints, mset...)
-			for s.Next() {
-				series := s.At()
-				labelValue := series.Labels().Get(name)
-				// Filter out empty value.
-				if labelValue == "" {
-					continue
-				}
-				labelValuesSet[labelValue] = struct{}{}
+		for _, matchers := range matcherSets {
+			vals, callWarnings, err = q.LabelValues(name, matchers...)
+			if err != nil {
+				return apiFuncResult{nil, &apiError{errorExec, err}, warnings, closer}
 			}
-			warnings = append(warnings, s.Warnings()...)
-			if err := s.Err(); err != nil {
-				return apiFuncResult{nil, &apiError{errorExec, err}, warnings, nil}
+			warnings = append(warnings, callWarnings...)
+			for _, val := range vals {
+				labelValuesSet[val] = struct{}{}
 			}
 		}
-
-		// Convert the map to an array.
-		vals = make([]string, 0, len(labelValuesSet))
-		for key := range labelValuesSet {
-			vals = append(vals, key)
-		}
-		sort.Strings(vals)
```

Comment:
I think we still need this :) looks like the tests are not covering this, could you add a test case covering this?

Reply:
I think we can definitely remove the `labelValuesSet` map, as this isn't even read anymore: 2e10b005f6bb1f77387f232227f5146438863f2f

There are actually unit tests for this function, which are also using matchers:
https://github.com/prometheus/prometheus/blob/06e53b7ecfca1350e73a6e75f789313ae01139d1/web/api/v1/api_test.go#L721-L775

But currently the results returned from `LabelValues()` in the unit tests are always sorted because the tests are using the `blockBaseQuerier`, which in its `LabelValues()` calls `q.index.SortedLabelValues()`: https://github.com/prometheus/prometheus/blob/3c3b76a02a95f6aa64d1d100f28793fa63ad9292/tsdb/querier.go#L87

Actually I don't know why the `blockBaseQuerier.LabelValues()` method would call `q.index.SortedLabelValues()` and I think it could just call `q.index.LabelValues()` instead, but I didn't add that so I'm hesitant to just change it. If I want to create a test which checks whether `api.labelValues()` sorts the list of values I'll have to make it use some alternative querier because as long the test is using the `blockBaseQuerier` the list of values would be sorted anyway.

Reply:
good catch, thx.

I added a unit test and a fix in `api.go`, without the fix in `api.go` the unit test was failing: 3a3290f8fccba86ee1b93144c033303b6d53ab2e